src/viam/sdk/common/proto_value.hpp

```diff
@@ -273,33 +297,56 @@ Struct map_to_struct(const ProtoStruct& m) {
     return s;
 }
 
-/// @brief ProtoValue RTTI type trait.
-/// This type trait is used to implement the ProtoValue::kind method which provides the type
-/// discriminator constant that is used in the value access API.
-/// A ProtoValue can only be constructed from types for which this trait is well formed.
+namespace proto_value_details {
+
 template <typename T>
-struct kind_t;
+struct kind {};
 
 template <>
-struct kind_t<std::nullptr_t> : std::integral_constant<int, 0> {};
+struct kind<std::nullptr_t> {
+    using type = std::integral_constant<int, 0>;
+};
 
 template <>
-struct kind_t<bool> : std::integral_constant<int, 1> {};
+struct kind<bool> {
+    using type = std::integral_constant<int, 1>;
+};
 
 template <>
-struct kind_t<int> : std::integral_constant<int, 2> {};
+struct kind<int> {
```

Comment:
It just occurred to me that proto's struct doesn't actually have an integer type. Just double. Do we want to be in the business of defining that conversion, or should we push that to users, since we cannot always know their intent.

Reply:
I've been wondering about this too--proto just has a `number` type which promotes everything to double. There are some gotchas in the unit tests about this:
https://github.com/lia-viam/viam-cpp-sdk/blob/feature/proto-value-visit/src/viam/sdk/tests/test_proto_value.cpp#L43

Reply:
I'm in favor of getting rid of the separate `int` and `double` types and replacing them with a `number` type which is just a `double` underneath, as this would match what's in proto
https://protobuf.dev/reference/protobuf/google.protobuf/#value

Deleting the conversion from `int` entirely is tempting, but maybe this will cause us headaches when migrating from `ProtoType`? The `variant` underlying that class also has distinct `int` and `double`, which is arguably a mistake for the same reason, but there will definitely be instances of `std::make_shared<ProtoType>(5)`

Reply:
OK, so several thoughts have come from this discussion:

- Does `ProtoValue` really need a `template` constructor, or should it just have constructors for each type. As written, you will just get a funky probably-link-time error because it can't instantiate it if you try to, say, construct a `ProtoValue` of some unrelated type.

- Should (some of the) `ProtoValue` constructors (templated or not) be `explicit`?

- I suspect that if we eliminate the ability to hold integers, then the one numeric constructor should take `double`, but allow implicit conversions (so, not explicit). That'd let `std::make_shared<ProtoType>(5)` still work, I think. What I want to avoid is confusing situations where by accepting `int` as a constructor argument, we allow things with conversion to `int` to be passed, but then those things become doubles internally.

Reply:
All good questions. My thoughts
- I may as well just spell out a bunch of non-`template` constructors for each type, although I think I would still want to have them defer to a private `template` constructor. Whatever amount of duplication is present it would still be fewer lines than all the explicit instantiations I'm writing
- I think yeah ideally we would want to follow the example of JSON libraries, where you have sort of easy ergonomic interfaces for building up objects like [this](https://www.boost.org/doc/libs/1_86_0/libs/json/doc/html/json/quick_look.html), which you get because the [constructors](https://www.boost.org/doc/libs/1_85_0/libs/json/doc/html/json/ref/boost__json__value/value.html) are not `explicit`.
- Yeah I think this makes sense and we will just be deferring to standard C++ language rules about type promotion/argument overloading

All that said, maybe I should make a separate PR to do
- have only `number` type
- define non-`template` public constructors

and then rebase this one off that one?

Reply:
Yes, let's go forward with what you have now where `int` is a thing, and then defer the whole `int` removal and any associated constructor fussing into a followup PR. I don't think you need to do the rebase part.

Reply:
OK cool so just to confirm should I still do the `kind_t` removal and changing to an `enum Kind` for this PR too?

Reply:
Yes, I think the `kind_t` removal is pretty trivial, and I'd really like to see the `default` case go away if possible, so I think the enum thing makes more sense in this review than in a new one.